# Employee Attrition Prediction using Machine Learning

## Exploratory Data Analysis (EDA)

---

### Author

**Aneesh Jose**

### Project Type

End-to-End Machine Learning Project

### Problem Domain

Human Resources Analytics

### Dataset

IBM HR Analytics Employee Attrition & Performance

---

## Project Goal

The purpose of this notebook is to perform a comprehensive exploratory data analysis on the IBM HR Analytics dataset to understand employee behavior, identify important predictors of attrition, evaluate data quality, and define the preprocessing strategy for the production machine learning pipeline.

---

## Expected Outcome

By the end of this notebook we will:

- Understand the dataset structure
- Identify missing values
- Detect duplicates
- Study feature distributions
- Analyze employee attrition
- Discover relationships between variables
- Produce actionable business insights
- Define preprocessing steps for model development

# Business Understanding

## Background

Employee attrition represents one of the major challenges faced by organizations.
High attrition leads to increased recruitment costs, productivity loss, reduced employee morale, and loss of organizational knowledge.

The Human Resources department aims to proactively identify employees who are likely to leave so that appropriate retention strategies can be implemented.

---

## Business Problem

Can we predict whether an employee is likely to leave the organization based on demographic, job-related, and performance-related attributes?

---

## Machine Learning Task

Binary Classification

Target Variable:

- Attrition

Classes:

- Yes
- No

# Dataset Information

The IBM HR Analytics Employee Attrition dataset contains employee demographic, educational, financial, and workplace-related information.

Each row represents one employee.

The dataset includes both numerical and categorical variables describing employee characteristics.

The objective is to predict whether an employee has left the organization.

In [70]:
import warnings
warnings.filterwarnings("ignore")

from pathlib import Path

import numpy as np
import pandas as pd

import plotly.express as px
import plotly.graph_objects as go

from plotly.subplots import make_subplots

from IPython.display import display

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 200)

import plotly.io as pio
pio.templates.default = "plotly_white"

In [71]:
DATA_PATH = Path("../dataset/employee_attrition.csv")

df = pd.read_csv(DATA_PATH)

print(f"Dataset Shape: {df.shape}")

df.head()

Dataset Shape: (1470, 35)


,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,2,Female,94,3,2,Sales Executive,4,Single,5993,19479,8,Y,Yes,11,3,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,3,Male,61,2,2,Research Scientist,2,Married,5130,24907,1,Y,No,23,4,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,4,Male,92,2,1,Laboratory Technician,3,Single,2090,2396,6,Y,Yes,15,3,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,4,Female,56,3,1,Research Scientist,3,Married,2909,23159,1,Y,Yes,11,3,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,1,Male,40,3,1,Laboratory Technician,2,Married,3468,16632,9,Y,No,12,3,4,80,1,6,3,3,2,2,2,2


#  Dataset Overview

## Objective

Before performing any preprocessing or visualization, it is essential to understand
the overall structure of the dataset.

This section provides a high-level overview of the dataset by examining its size,
feature composition, memory usage, and overall quality.

Understanding these characteristics helps determine the complexity of the dataset
and guides the preprocessing strategy in the later stages of the machine learning
pipeline.

In [72]:
print("=" * 80)
print("DATASET SHAPE")
print("=" * 80)

print(f"Rows    : {df.shape[0]:,}")
print(f"Columns : {df.shape[1]}")

print("=" * 80)

df.head()

DATASET SHAPE
Rows    : 1,470
Columns : 35


,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,2,Female,94,3,2,Sales Executive,4,Single,5993,19479,8,Y,Yes,11,3,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,3,Male,61,2,2,Research Scientist,2,Married,5130,24907,1,Y,No,23,4,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,4,Male,92,2,1,Laboratory Technician,3,Single,2090,2396,6,Y,Yes,15,3,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,4,Female,56,3,1,Research Scientist,3,Married,2909,23159,1,Y,Yes,11,3,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,1,Male,40,3,1,Laboratory Technician,2,Married,3468,16632,9,Y,No,12,3,4,80,1,6,3,3,2,2,2,2


#  Dataset Information

## Objective

This section examines the metadata of the dataset.

Specifically, we analyze:

- Feature names
- Data types
- Number of non-null observations
- Memory usage

These insights help identify categorical variables,
numerical variables, and possible data quality issues.

In [73]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   Age                       1470 non-null   int64
 1   Attrition                 1470 non-null   str  
 2   BusinessTravel            1470 non-null   str  
 3   DailyRate                 1470 non-null   int64
 4   Department                1470 non-null   str  
 5   DistanceFromHome          1470 non-null   int64
 6   Education                 1470 non-null   int64
 7   EducationField            1470 non-null   str  
 8   EmployeeCount             1470 non-null   int64
 9   EmployeeNumber            1470 non-null   int64
 10  EnvironmentSatisfaction   1470 non-null   int64
 11  Gender                    1470 non-null   str  
 12  HourlyRate                1470 non-null   int64
 13  JobInvolvement            1470 non-null   int64
 14  JobLevel                  1470 non-null   int64
 15

# Dataset Summary Dashboard

Instead of inspecting each statistic separately, this dashboard summarizes the
overall characteristics of the dataset in one place.

The dashboard provides an instant understanding of:

- Dataset dimensions
- Missing values
- Duplicate rows
- Numerical features
- Categorical features
- Memory usage

In [74]:
rows = df.shape[0]
columns = df.shape[1]

missing_values = df.isnull().sum().sum()

duplicate_rows = df.duplicated().sum()

numerical_features = len(
    df.select_dtypes(include=np.number).columns
)

categorical_features = len(
    df.select_dtypes(exclude=np.number).columns
)

memory_usage = round(
    df.memory_usage(deep=True).sum() / 1024**2,
    2
)

In [75]:
fig = go.Figure()

fig.add_trace(
    go.Indicator(
        mode="number",
        value=rows,
        title={"text": "Rows"},
        domain={"x": [0.0, 0.18], "y": [0.55, 1]}
    )
)

fig.add_trace(
    go.Indicator(
        mode="number",
        value=columns,
        title={"text": "Columns"},
        domain={"x": [0.20, 0.38], "y": [0.55, 1]}
    )
)

fig.add_trace(
    go.Indicator(
        mode="number",
        value=numerical_features,
        title={"text": "Numerical"},
        domain={"x": [0.40, 0.58], "y": [0.55, 1]}
    )
)

fig.add_trace(
    go.Indicator(
        mode="number",
        value=categorical_features,
        title={"text": "Categorical"},
        domain={"x": [0.60, 0.78], "y": [0.55, 1]}
    )
)

fig.add_trace(
    go.Indicator(
        mode="number",
        value=missing_values,
        title={"text": "Missing"},
        domain={"x": [0.80, 1], "y": [0.55, 1]}
    )
)

fig.add_trace(
    go.Indicator(
        mode="number",
        value=duplicate_rows,
        title={"text": "Duplicates"},
        domain={"x": [0.20, 0.45], "y": [0, 0.4]}
    )
)

fig.add_trace(
    go.Indicator(
        mode="number",
        value=memory_usage,
        number={"suffix": " MB"},
        title={"text": "Memory"},
        domain={"x": [0.55, 0.80], "y": [0, 0.4]}
    )
)

fig.update_layout(
    title="Dataset Summary Dashboard",
    height=450
)

fig.show()

In [76]:
display(df.head())

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,2,Female,94,3,2,Sales Executive,4,Single,5993,19479,8,Y,Yes,11,3,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,3,Male,61,2,2,Research Scientist,2,Married,5130,24907,1,Y,No,23,4,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,4,Male,92,2,1,Laboratory Technician,3,Single,2090,2396,6,Y,Yes,15,3,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,4,Female,56,3,1,Research Scientist,3,Married,2909,23159,1,Y,Yes,11,3,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,1,Male,40,3,1,Laboratory Technician,2,Married,3468,16632,9,Y,No,12,3,4,80,1,6,3,3,2,2,2,2


In [77]:
display(df.sample(10, random_state=42))

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
1041,28,No,Travel_Rarely,866,Sales,5,3,Medical,1,1469,4,Male,84,3,2,Sales Executive,1,Single,8463,23490,0,Y,No,18,3,4,80,0,6,4,3,5,4,1,3
184,53,No,Travel_Rarely,1084,Research & Development,13,2,Medical,1,250,4,Female,57,4,2,Manufacturing Director,1,Divorced,4450,26250,1,Y,No,11,3,3,80,2,5,3,3,4,2,1,3
1222,24,Yes,Travel_Rarely,240,Human Resources,22,1,Human Resources,1,1714,4,Male,58,1,1,Human Resources,3,Married,1555,11585,1,Y,No,11,3,3,80,1,1,2,3,1,0,0,0
67,45,No,Travel_Rarely,1339,Research & Development,7,3,Life Sciences,1,86,2,Male,59,3,3,Research Scientist,1,Divorced,9724,18787,2,Y,No,17,3,3,80,1,25,2,3,1,0,0,0
220,36,No,Travel_Rarely,1396,Research & Development,5,2,Life Sciences,1,304,4,Male,62,3,2,Laboratory Technician,2,Single,5914,9945,8,Y,No,16,3,4,80,0,16,3,4,13,11,3,7
494,34,No,Travel_Rarely,204,Sales,14,3,Technical Degree,1,666,3,Female,31,3,1,Sales Representative,3,Divorced,2579,2912,1,Y,Yes,18,3,4,80,2,8,3,3,8,2,0,6
430,35,No,Travel_Rarely,144,Research & Development,22,3,Life Sciences,1,577,4,Male,46,1,1,Laboratory Technician,3,Single,4230,19225,0,Y,No,15,3,3,80,0,6,2,3,5,4,4,3
240,39,No,Travel_Rarely,1431,Research & Development,1,4,Medical,1,332,3,Female,96,3,1,Laboratory Technician,3,Divorced,2232,15417,7,Y,No,14,3,3,80,3,7,1,3,3,2,1,2
218,45,No,Non-Travel,1052,Sales,6,3,Medical,1,302,4,Female,57,2,3,Sales Executive,4,Single,8865,16840,6,Y,No,12,3,4,80,0,23,2,3,19,7,12,8
49,35,No,Travel_Rarely,1229,Research & Development,8,1,Life Sciences,1,63,4,Male,36,4,1,Laboratory Technician,4,Married,2269,4892,1,Y,No,19,3,4,80,0,1,2,3,1,0,0,1


# Feature Classification

## Objective

Machine learning preprocessing depends heavily on feature types.

This section automatically separates the features into:

- Numerical Features
- Categorical Features

These lists will later be used to build the preprocessing pipeline using
ColumnTransformer.

In [78]:
numerical_columns = (
    df
    .select_dtypes(include=np.number)
    .columns
    .tolist()
)

categorical_columns = (
    df
    .select_dtypes(exclude=np.number)
    .columns
    .tolist()
)

In [79]:
print("Numerical Features")

numerical_columns

Numerical Features


['Age',
 'DailyRate',
 'DistanceFromHome',
 'Education',
 'EmployeeCount',
 'EmployeeNumber',
 'EnvironmentSatisfaction',
 'HourlyRate',
 'JobInvolvement',
 'JobLevel',
 'JobSatisfaction',
 'MonthlyIncome',
 'MonthlyRate',
 'NumCompaniesWorked',
 'PercentSalaryHike',
 'PerformanceRating',
 'RelationshipSatisfaction',
 'StandardHours',
 'StockOptionLevel',
 'TotalWorkingYears',
 'TrainingTimesLastYear',
 'WorkLifeBalance',
 'YearsAtCompany',
 'YearsInCurrentRole',
 'YearsSinceLastPromotion',
 'YearsWithCurrManager']

In [80]:
numerical_df = pd.DataFrame(
    {
        "Numerical Features": numerical_columns
    }
)

numerical_df

,Numerical Features
0,Age
1,DailyRate
2,DistanceFromHome
3,Education
4,EmployeeCount
5,EmployeeNumber
6,EnvironmentSatisfaction
7,HourlyRate
8,JobInvolvement
9,JobLevel


In [81]:
categorical_df = pd.DataFrame(
    {
        "Categorical Features": categorical_columns
    }
)

categorical_df

,Categorical Features
0,Attrition
1,BusinessTravel
2,Department
3,EducationField
4,Gender
5,JobRole
6,MaritalStatus
7,Over18
8,OverTime


# Key Findings

## Dataset Structure

The IBM HR Analytics dataset contains **1,470 employee records** and **35 features**, where each row represents an individual employee and each column describes a demographic, professional, or workplace-related characteristic.

The dataset size is appropriate for developing and evaluating a supervised machine learning classification model.

---

## Data Quality

The dataset is complete and well-structured, with **no missing values** across any of the 35 features.

No inconsistencies were identified in the initial inspection of the dataset.

---

## Feature Composition

The dataset contains a combination of:

- **26 Numerical Features**
- **9 Categorical Features**

This mixture of numerical and categorical variables makes the dataset suitable for applying preprocessing techniques such as feature encoding and numerical scaling before model training.

---

## Memory Usage

The dataset occupies approximately **510.8 KB** of memory, making it lightweight and suitable for efficient in-memory processing without requiring distributed computing techniques.

---

## Initial Assessment

The dataset appears clean and requires minimal data cleaning before model development.

The next phase of the analysis will focus on identifying duplicate records, constant-value features, target variable distribution, feature relationships, and potential preprocessing requirements.

# Data Quality Assessment

## Objective

Before analyzing relationships between variables, it is essential to assess the overall quality of the dataset.

A high-quality dataset improves model performance, reduces preprocessing complexity, and minimizes the risk of introducing bias during training.

In this section, we will evaluate:

- Duplicate records
- Constant-value features
- Feature uniqueness
- Data types
- Categorical feature cardinality

The findings from this analysis will help determine which features should be retained, transformed, or removed before model development.

In [82]:
duplicate_count = df.duplicated().sum()

print(f"Number of Duplicate Records: {duplicate_count}")

Number of Duplicate Records: 0


In [83]:
import plotly.graph_objects as go

fig = go.Figure(
    go.Indicator(
        mode="gauge+number",
        value=duplicate_count,
        title={"text": "Duplicate Records"},
        gauge={
            "axis": {"range": [0, max(1, duplicate_count + 5)]},
            "bar": {"color": "royalblue"}
        }
    )
)

fig.update_layout(height=400)
fig.show()

# Constant Feature Analysis

## Objective

Features with only one unique value contain no information that helps distinguish between classes.

Such variables contribute no predictive power and should be removed before model training.

In [84]:
constant_columns = [
    col for col in df.columns
    if df[col].nunique() == 1
]

constant_columns

['EmployeeCount', 'Over18', 'StandardHours']

In [85]:
constant_df = pd.DataFrame(
    {
        "Constant Features": constant_columns
    }
)

constant_df

,Constant Features
0,EmployeeCount
1,Over18
2,StandardHours


# Key Findings

Three features contain only a single unique value across all employee records:

- EmployeeCount
- Over18
- StandardHours

Since these variables exhibit zero variance, they do not contribute any useful information for distinguishing employees who leave the organization from those who remain.

Including such features would unnecessarily increase model complexity without improving predictive performance.
# Machine Learning Recommendation

The following features will be removed during preprocessing:

- EmployeeCount
- Over18
- StandardHours

These variables provide no predictive value and are considered zero-variance features.

# Feature Cardinality Analysis

## Objective

Feature cardinality refers to the number of unique values present in a feature.

Understanding cardinality helps identify:

- Identifier columns
- Constant features
- Binary variables
- Low-cardinality categorical variables
- High-cardinality categorical variables

This information guides feature engineering and encoding strategies.

In [86]:
unique_values = pd.DataFrame({
    "Feature": df.columns,
    "Data Type": df.dtypes.astype(str),
    "Unique Values": [df[col].nunique() for col in df.columns]
})

unique_values.sort_values(
    by="Unique Values",
    inplace=True
)

unique_values.reset_index(drop=True)

,Feature,Data Type,Unique Values
0,EmployeeCount,int64,1
1,Over18,str,1
2,StandardHours,int64,1
3,Attrition,str,2
4,OverTime,str,2
5,PerformanceRating,int64,2
6,Gender,str,2
7,BusinessTravel,str,3
8,Department,str,3
9,MaritalStatus,str,3


In [87]:
fig = px.bar(
    unique_values,
    x="Feature",
    y="Unique Values",
    color="Data Type",
    title="Number of Unique Values per Feature"
)

fig.update_layout(
    xaxis_tickangle=-60,
    height=650
)

fig.show()

# Key Findings

The feature cardinality analysis reveals different categories of variables:

- Constant-value features with a single unique value.
- Binary categorical variables such as Attrition, Gender, and OverTime.
- Low-cardinality categorical features including Department and JobRole.
- High-cardinality numerical variables such as EmployeeNumber and MonthlyIncome.

EmployeeNumber contains a unique identifier for each employee and does not represent meaningful business information for prediction.
# Machine Learning Recommendation

The EmployeeNumber feature will be excluded from model training because it serves solely as a unique identifier and does not contribute to predictive performance.

Binary and low-cardinality categorical variables will be encoded during preprocessing using appropriate encoding techniques.

# Target Variable Analysis

## Objective

The target variable is the feature that the machine learning model aims to predict.

In this project, the target variable is **Attrition**, which indicates whether an employee has left the organization.

Analyzing the target variable helps us:

- Understand the distribution of employee attrition.
- Identify potential class imbalance.
- Estimate the baseline difficulty of the classification problem.
- Determine whether special techniques such as class weighting or resampling may be required during model training.

This analysis forms the foundation for selecting suitable evaluation metrics and machine learning algorithms.

In [88]:
print("Target Variable :", df["Attrition"].name)
print()

print(df["Attrition"].value_counts())

Target Variable : Attrition

Attrition
No     1233
Yes     237
Name: count, dtype: int64


In [89]:
attrition_percentage = (
    df["Attrition"]
    .value_counts(normalize=True)
    .mul(100)
    .round(2)
)

attrition_percentage

Attrition
No     83.88
Yes    16.12
Name: proportion, dtype: float64

In [90]:
attrition_summary = pd.DataFrame({
    "Count": df["Attrition"].value_counts(),
    "Percentage": attrition_percentage
})

attrition_summary

,Count,Percentage
Attrition,,
No,1233,83.88
Yes,237,16.12


In [91]:
fig = px.bar(
    attrition_summary.reset_index(),
    x="Attrition",
    y="Count",
    color="Attrition",
    text="Count",
    title="Employee Attrition Distribution"
)

fig.update_traces(textposition="outside")

fig.update_layout(
    showlegend=False,
    height=500
)

fig.show()

In [92]:
fig = px.pie(
    attrition_summary.reset_index(),
    names="Attrition",
    values="Count",
    hole=0.45,
    title="Percentage Distribution of Employee Attrition"
)

fig.update_traces(
    textposition="inside",
    textinfo="percent+label"
)

fig.show()

In [93]:
attrition_rate = (
    df["Attrition"]
    .value_counts(normalize=True)["Yes"] * 100
)

fig = go.Figure()

fig.add_trace(
    go.Indicator(
        mode="gauge+number",
        value=attrition_rate,
        number={"suffix": "%"},
        title={"text": "Overall Attrition Rate"},
        gauge={
            "axis": {"range": [0, 100]},
            "bar": {"color": "crimson"}
        }
    )
)

fig.update_layout(height=450)

fig.show()

# Key Findings

The target variable consists of two classes:

- **No** – Employees who remained in the organization.
- **Yes** – Employees who left the organization.

Approximately **84%** of employees remained with the company, while **16%** left.

This indicates that employee attrition is relatively infrequent compared to employee retention.

# Business Interpretation

The organization demonstrates a relatively high employee retention rate, with only a small proportion of employees leaving.

Although the attrition rate is comparatively low, employee turnover remains an important business concern because replacing experienced employees involves significant recruitment, onboarding, and training costs.

Understanding the factors associated with employee attrition can help the Human Resources department develop targeted retention strategies, improve employee satisfaction, and reduce organizational costs.

# Machine Learning Recommendation

The target variable exhibits a noticeable class imbalance, with the majority class ("No") significantly outnumbering the minority class ("Yes").

Although the imbalance is not severe, relying solely on accuracy could produce misleading results because a model predicting only the majority class would still achieve high accuracy.

Therefore, model evaluation will emphasize metrics such as:

- Precision
- Recall
- F1-Score
- ROC-AUC

During model experimentation, techniques such as class weighting or SMOTE may also be evaluated if necessary.


In [94]:
employees_left = (df["Attrition"] == "Yes").sum()
employees_retained = (df["Attrition"] == "No").sum()

fig = go.Figure()

fig.add_trace(go.Indicator(
    mode="number",
    value=employees_left,
    title={"text": "Employees Left"},
    domain={"x": [0, 0.45], "y": [0, 1]}
))

fig.add_trace(go.Indicator(
    mode="number",
    value=employees_retained,
    title={"text": "Employees Retained"},
    domain={"x": [0.55, 1], "y": [0, 1]}
))

fig.update_layout(
    title="Employee Retention Summary",
    height=300
)

fig.show()

# Univariate Analysis of Numerical Features

## Objective

The objective of this section is to understand the distribution of each numerical feature independently.

Univariate analysis helps identify:

- Distribution patterns
- Central tendency
- Spread of the data
- Presence of skewness
- Potential outliers
- Features that may require scaling or transformation

These insights guide preprocessing decisions and help determine the suitability of different machine learning algorithms.

In [95]:
numerical_columns = df.select_dtypes(include=["int64", "float64"]).columns.tolist()

print(f"Number of Numerical Features: {len(numerical_columns)}")

numerical_columns

Number of Numerical Features: 26


['Age',
 'DailyRate',
 'DistanceFromHome',
 'Education',
 'EmployeeCount',
 'EmployeeNumber',
 'EnvironmentSatisfaction',
 'HourlyRate',
 'JobInvolvement',
 'JobLevel',
 'JobSatisfaction',
 'MonthlyIncome',
 'MonthlyRate',
 'NumCompaniesWorked',
 'PercentSalaryHike',
 'PerformanceRating',
 'RelationshipSatisfaction',
 'StandardHours',
 'StockOptionLevel',
 'TotalWorkingYears',
 'TrainingTimesLastYear',
 'WorkLifeBalance',
 'YearsAtCompany',
 'YearsInCurrentRole',
 'YearsSinceLastPromotion',
 'YearsWithCurrManager']

In [96]:
numerical_summary = (
    df[numerical_columns]
    .describe()
    .T
    .round(2)
)

numerical_summary

,count,mean,std,min,25%,50%,75%,max
Age,1470.0,36.92,9.14,18.0,30.00,36.0,43.00,60.0
DailyRate,1470.0,802.49,403.51,102.0,465.00,802.0,1157.00,1499.0
DistanceFromHome,1470.0,9.19,8.11,1.0,2.00,7.0,14.00,29.0
Education,1470.0,2.91,1.02,1.0,2.00,3.0,4.00,5.0
EmployeeCount,1470.0,1.00,0.00,1.0,1.00,1.0,1.00,1.0
EmployeeNumber,1470.0,1024.87,602.02,1.0,491.25,1020.5,1555.75,2068.0
EnvironmentSatisfaction,1470.0,2.72,1.09,1.0,2.00,3.0,4.00,4.0
HourlyRate,1470.0,65.89,20.33,30.0,48.00,66.0,83.75,100.0
JobInvolvement,1470.0,2.73,0.71,1.0,2.00,3.0,3.00,4.0
JobLevel,1470.0,2.06,1.11,1.0,1.00,2.0,3.00,5.0


In [97]:
from plotly.subplots import make_subplots
import plotly.graph_objects as go

rows = 9
cols = 3

fig = make_subplots(
    rows=rows,
    cols=cols,
    subplot_titles=numerical_columns
)

for index, column in enumerate(numerical_columns):

    r = index // cols + 1
    c = index % cols + 1

    fig.add_trace(
        go.Histogram(
            x=df[column],
            showlegend=False
        ),
        row=r,
        col=c
    )

fig.update_layout(
    height=2600,
    title="Distribution of Numerical Features"
)

fig.show()

In [98]:
fig = make_subplots(
    rows=9,
    cols=3,
    subplot_titles=numerical_columns
)

for index, column in enumerate(numerical_columns):

    r = index // 3 + 1
    c = index % 3 + 1

    fig.add_trace(
        go.Box(
            y=df[column],
            showlegend=False,
            name=column
        ),
        row=r,
        col=c
    )

fig.update_layout(
    height=2600,
    title="Outlier Detection using Boxplots"
)

fig.show()

### Age

#### Objective

To understand the age distribution of employees and identify whether the workforce is concentrated within a particular age group.

#### Key Findings

Most employees are concentrated in the early and middle stages of their careers, with relatively fewer employees at the extreme ends of the age spectrum.

#### Business Interpretation

A workforce dominated by mid-career professionals may have different retention patterns compared with organizations employing predominantly younger or older employees.

#### ML Recommendation

No transformation is required.

Age will be retained as a numerical feature and scaled if required by the selected machine learning algorithm.

### Monthly Income

#### Key Findings

Monthly income exhibits a right-skewed distribution, with a relatively small number of employees earning substantially higher salaries.

#### Business Interpretation

Salary differences may reflect variations in experience, job level, and organizational responsibilities.

Income is expected to be an important predictor of employee attrition.

#### ML Recommendation

Scaling will be considered during preprocessing.

Potential interaction with JobLevel and TotalWorkingYears should be explored during bivariate analysis.

### Total Working Years

#### Key Findings

Employee work experience varies considerably across the organization.

While many employees possess moderate experience, a smaller group has extensive professional experience.

#### Business Interpretation

Experience may influence employee retention through career stability, compensation, and organizational commitment.

#### ML Recommendation

Retain for further relationship analysis with the target variable.

# Overall Findings – Numerical Features

The numerical variables display diverse distributions and levels of variability.

Several variables, including Monthly Income and Total Working Years, exhibit right-skewed distributions, while others are more evenly distributed.

Boxplot analysis identifies potential outliers in compensation- and experience-related variables. However, these observations likely represent genuine employee characteristics rather than data entry errors.

Consequently, outliers will be retained unless subsequent analysis indicates that they negatively affect model performance.

# Machine Learning Decisions

Based on the univariate numerical analysis, the following preprocessing strategy is recommended:

- Remove constant-value features identified earlier.
- Retain naturally occurring outliers unless justified otherwise.
- Apply feature scaling for algorithms sensitive to feature magnitude (e.g., Logistic Regression, KNN, SVM).
- Evaluate skewed variables during feature engineering but avoid unnecessary transformations without evidence of improved model performance.

#  Univariate Analysis of Categorical Features

## Objective

Categorical variables describe employee characteristics such as department, job role, gender, business travel frequency, and marital status.

Understanding the distribution of these variables helps identify workforce composition and provides valuable business context before examining their relationship with employee attrition.

The objectives of this section are to:

- Examine the frequency distribution of categorical features.
- Identify dominant employee groups.
- Detect categories with very few observations.
- Prepare for feature encoding in the machine learning pipeline.

In [99]:
categorical_columns = df.select_dtypes(include="object").columns.tolist()

print(f"Categorical Features : {len(categorical_columns)}")

categorical_columns

Categorical Features : 9


['Attrition',
 'BusinessTravel',
 'Department',
 'EducationField',
 'Gender',
 'JobRole',
 'MaritalStatus',
 'Over18',
 'OverTime']

In [100]:
summary = pd.DataFrame({
    "Feature": categorical_columns,
    "Unique Categories":
    [df[col].nunique() for col in categorical_columns]
})

summary

,Feature,Unique Categories
0,Attrition,2
1,BusinessTravel,3
2,Department,3
3,EducationField,6
4,Gender,2
5,JobRole,9
6,MaritalStatus,3
7,Over18,1
8,OverTime,2


In [101]:
for column in categorical_columns:

    print("="*70)

    print(column)

    print("="*70)

    display(df[column].value_counts())

Attrition


Attrition
No     1233
Yes     237
Name: count, dtype: int64

BusinessTravel


BusinessTravel
Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: count, dtype: int64

Department


Department
Research & Development    961
Sales                     446
Human Resources            63
Name: count, dtype: int64

EducationField


EducationField
Life Sciences       606
Medical             464
Marketing           159
Technical Degree    132
Other                82
Human Resources      27
Name: count, dtype: int64

Gender


Gender
Male      882
Female    588
Name: count, dtype: int64

JobRole


JobRole
Sales Executive              326
Research Scientist           292
Laboratory Technician        259
Manufacturing Director       145
Healthcare Representative    131
Manager                      102
Sales Representative          83
Research Director             80
Human Resources               52
Name: count, dtype: int64

MaritalStatus


MaritalStatus
Married     673
Single      470
Divorced    327
Name: count, dtype: int64

Over18


Over18
Y    1470
Name: count, dtype: int64

OverTime


OverTime
No     1054
Yes     416
Name: count, dtype: int64

In [102]:
for column in categorical_columns:

    counts = (
        df[column]
        .value_counts()
        .reset_index()
    )

    counts.columns = [column, "Count"]

    fig = px.bar(
        counts,
        x=column,
        y="Count",
        color=column,
        text="Count",
        title=f"{column} Distribution"
    )

    fig.update_traces(textposition="outside")

    fig.show()

# Overall Findings – Categorical Features

The categorical variables describe important aspects of employee demographics and workplace characteristics.

Most variables contain a manageable number of categories, making them suitable for one-hot encoding during preprocessing.

Features such as Department, JobRole, BusinessTravel, and OverTime appear particularly relevant for explaining employee attrition and will be examined further during bivariate analysis.

# Machine Learning Recommendation

Based on the categorical feature analysis, the following preprocessing strategy is proposed:

- Remove constant categorical features such as Over18.
- Apply One-Hot Encoding to nominal categorical variables.
- Preserve all informative categorical features.
- Evaluate the relationship between categorical variables and the target variable before feature selection.

# Bivariate Analysis

## Objective

The objective of this section is to investigate the relationship between individual employee characteristics and the target variable (**Attrition**).

Unlike univariate analysis, which examines each feature independently, bivariate analysis explores how different employee attributes influence the likelihood of attrition.

The findings from this analysis will help identify the most influential predictors and generate actionable business recommendations for improving employee retention.

The analysis focuses on answering key business questions related to workforce behavior and employee turnover.

In [103]:
overtime_attrition = pd.crosstab(
    df["OverTime"],
    df["Attrition"]
)

overtime_attrition

Attrition,No,Yes
OverTime,,
No,944,110
Yes,289,127


In [104]:
overtime_percentage = pd.crosstab(
    df["OverTime"],
    df["Attrition"],
    normalize="index"
) * 100

overtime_percentage.round(2)

Attrition,No,Yes
OverTime,,
No,89.56,10.44
Yes,69.47,30.53


In [105]:
fig = px.bar(
    overtime_percentage.reset_index(),
    x="OverTime",
    y=["No","Yes"],
    barmode="stack",
    text_auto=".1f",
    title="Employee Attrition by Overtime Status"
)

fig.update_layout(
    yaxis_title="Percentage"
)

fig.show()

## Key Findings

Employees who work overtime exhibit a noticeably higher attrition rate compared with employees who do not work overtime.

The difference is substantial, suggesting that overtime is strongly associated with employee turnover.

## Business Interpretation

Extended working hours may contribute to employee fatigue, increased stress, and reduced work-life balance.

Organizations should monitor overtime practices and evaluate whether excessive workloads are contributing to employee resignations.

## Machine Learning Recommendation

OverTime appears to be a highly informative feature and should be retained for model training.

This feature is expected to receive high importance scores in tree-based models.

In [106]:
department_attrition = pd.crosstab(
    df["Department"],
    df["Attrition"],
    normalize="index"
) * 100

department_attrition.round(2)

Attrition,No,Yes
Department,,
Human Resources,80.95,19.05
Research & Development,86.16,13.84
Sales,79.37,20.63


In [107]:
fig = px.bar(
    department_attrition.reset_index(),
    x="Department",
    y=["No","Yes"],
    barmode="stack",
    title="Attrition Rate by Department",
    text_auto=".1f"
)

fig.show()

Research & Development employs the largest workforce, while Human Resources has the smallest.

Although departments differ in size, attrition rates should be compared using percentages rather than raw employee counts to ensure a fair comparison.

Departments exhibiting higher attrition percentages may require targeted retention initiatives.

In [108]:
job_sat = pd.crosstab(
    df["JobSatisfaction"],
    df["Attrition"],
    normalize="index"
) * 100

job_sat.round(2)

Attrition,No,Yes
JobSatisfaction,,
1,77.16,22.84
2,83.57,16.43
3,83.48,16.52
4,88.67,11.33


In [109]:
fig = px.line(
    job_sat.reset_index(),
    x="JobSatisfaction",
    y="Yes",
    markers=True,
    title="Attrition Rate by Job Satisfaction"
)

fig.show()

Lower job satisfaction levels are generally associated with higher employee attrition.

Improving workplace satisfaction may therefore contribute to better employee retention.

In [110]:
fig = px.box(
    df,
    x="Attrition",
    y="MonthlyIncome",
    color="Attrition",
    title="Monthly Income vs Attrition"
)

fig.show()

Employees who leave the organization tend to exhibit different salary distributions than employees who remain.

Compensation alone may not explain employee turnover; however, salary should be evaluated alongside job level, work experience, and overtime.

In [111]:
fig = px.histogram(
    df,
    x="Age",
    color="Attrition",
    marginal="box",
    nbins=25,
    title="Age Distribution by Attrition"
)

fig.show()

Employee attrition appears more common among younger employees.

This pattern may reflect career mobility, early-career exploration, or greater willingness to change employers.

In [112]:
wlb = pd.crosstab(
    df["WorkLifeBalance"],
    df["Attrition"],
    normalize="index"
) * 100

wlb.round(2)

Attrition,No,Yes
WorkLifeBalance,,
1,68.75,31.25
2,83.14,16.86
3,85.78,14.22
4,82.35,17.65


In [113]:
fig = px.bar(
    wlb.reset_index(),
    x="WorkLifeBalance",
    y=["No","Yes"],
    barmode="stack",
    title="Work-Life Balance vs Attrition",
    text_auto=".1f"
)

fig.show()

Employees reporting poorer work-life balance exhibit higher attrition rates.

This suggests that flexible work policies and workload management could improve employee retention.

In [114]:
travel = pd.crosstab(
    df["BusinessTravel"],
    df["Attrition"],
    normalize="index"
) * 100

travel.round(2)

Attrition,No,Yes
BusinessTravel,,
Non-Travel,92.00,8.00
Travel_Frequently,75.09,24.91
Travel_Rarely,85.04,14.96


In [115]:
fig = px.bar(
    travel.reset_index(),
    x="BusinessTravel",
    y=["No","Yes"],
    barmode="stack",
    title="Business Travel vs Attrition"
)

fig.show()

Employees who travel frequently may exhibit higher attrition compared with employees who travel rarely or not at all.

Travel frequency should therefore be considered an important workplace factor.

# Preliminary Findings

Based on the exploratory analysis, the following features appear to have the strongest relationship with employee attrition:

- OverTime
- JobSatisfaction
- MonthlyIncome
- WorkLifeBalance
- TotalWorkingYears
- JobRole
- BusinessTravel
- Age

These observations are exploratory and will be validated during model training using feature importance analysis and model interpretation techniques.

# Correlation Analysis

## Objective

Correlation analysis measures the strength and direction of the linear relationship between numerical variables.

Understanding these relationships helps us:

- Identify highly correlated features.
- Detect potential multicollinearity.
- Understand how employee-related characteristics interact.
- Guide feature selection and engineering.

It is important to note that correlation does not imply causation. The analysis is used as an exploratory tool to identify patterns that may be useful during model development.

In [116]:
corr_matrix = df[numerical_columns].corr()

corr_matrix.head()

,Age,DailyRate,DistanceFromHome,Education,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,HourlyRate,JobInvolvement,JobLevel,JobSatisfaction,MonthlyIncome,MonthlyRate,NumCompaniesWorked,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
Age,1.000000,0.010661,-0.001686,0.208034,NaN,-0.010145,0.010146,0.024287,0.029820,0.509604,-0.004892,0.497855,0.028051,0.299635,0.003634,0.001904,0.053535,NaN,0.037510,0.680381,-0.019621,-0.021490,0.311309,0.212901,0.216513,0.202089
DailyRate,0.010661,1.000000,-0.004985,-0.016806,NaN,-0.050990,0.018355,0.023381,0.046135,0.002966,0.030571,0.007707,-0.032182,0.038153,0.022704,0.000473,0.007846,NaN,0.042143,0.014515,0.002453,-0.037848,-0.034055,0.009932,-0.033229,-0.026363
DistanceFromHome,-0.001686,-0.004985,1.000000,0.021042,NaN,0.032916,-0.016075,0.031131,0.008783,0.005303,-0.003669,-0.017014,0.027473,-0.029251,0.040235,0.027110,0.006557,NaN,0.044872,0.004628,-0.036942,-0.026556,0.009508,0.018845,0.010029,0.014406
Education,0.208034,-0.016806,0.021042,1.000000,NaN,0.042070,-0.027128,0.016775,0.042438,0.101589,-0.011296,0.094961,-0.026084,0.126317,-0.011111,-0.024539,-0.009118,NaN,0.018422,0.148280,-0.025100,0.009819,0.069114,0.060236,0.054254,0.069065
EmployeeCount,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [117]:
import plotly.express as px

fig = px.imshow(
    corr_matrix,
    text_auto=".2f",
    aspect="auto",
    color_continuous_scale="RdBu_r",
    zmin=-1,
    zmax=1,
    title="Correlation Matrix of Numerical Features"
)

fig.update_layout(
    height=900,
    width=900
)

fig.show()

In [118]:
corr_pairs = (
    corr_matrix
    .where(np.triu(np.ones(corr_matrix.shape), k=1).astype(bool))
    .stack()
    .reset_index()
)

corr_pairs.columns = [
    "Feature 1",
    "Feature 2",
    "Correlation"
]

strong_corr = (
    corr_pairs[
        abs(corr_pairs["Correlation"]) >= 0.70
    ]
    .sort_values(
        by="Correlation",
        ascending=False
    )
)

strong_corr

,Feature 1,Feature 2,Correlation
245,JobLevel,MonthlyIncome,0.950300
253,JobLevel,TotalWorkingYears,0.782208
379,PercentSalaryHike,PerformanceRating,0.773550
305,MonthlyIncome,TotalWorkingYears,0.772893
597,YearsAtCompany,YearsWithCurrManager,0.769212
595,YearsAtCompany,YearsInCurrentRole,0.758754
623,YearsInCurrentRole,YearsWithCurrManager,0.714365


# Key Findings

Most numerical variables exhibit weak to moderate correlations, suggesting that each feature contributes unique information.

A small number of features show stronger positive relationships, particularly those related to employee experience and tenure.

No evidence of widespread multicollinearity is observed based on the initial correlation analysis.

# Machine Learning Recommendation

Since severe multicollinearity is not observed, the current feature set can be retained for model experimentation.

If future models such as Logistic Regression exhibit instability, additional techniques such as Variance Inflation Factor (VIF) analysis may be considered.

# Multivariate Insights

## Objective

The purpose of multivariate analysis is to combine findings from previous sections and identify the employee characteristics that appear to influence attrition collectively.

Rather than evaluating variables independently, this section summarizes interactions among demographic, workplace, and performance-related features.

In [119]:
insights = pd.DataFrame({
    "Feature": [
        "OverTime",
        "JobSatisfaction",
        "MonthlyIncome",
        "WorkLifeBalance",
        "BusinessTravel",
        "Age",
        "DistanceFromHome",
        "TotalWorkingYears"
    ],
    "Observed Relationship": [
        "Strong",
        "Moderate",
        "Moderate",
        "Moderate",
        "Moderate",
        "Moderate",
        "Weak to Moderate",
        "Moderate"
    ],
    "Business Impact": [
        "High",
        "High",
        "Medium",
        "High",
        "Medium",
        "Medium",
        "Medium",
        "High"
    ]
})

insights

,Feature,Observed Relationship,Business Impact
0,OverTime,Strong,High
1,JobSatisfaction,Moderate,High
2,MonthlyIncome,Moderate,Medium
3,WorkLifeBalance,Moderate,High
4,BusinessTravel,Moderate,Medium
5,Age,Moderate,Medium
6,DistanceFromHome,Weak to Moderate,Medium
7,TotalWorkingYears,Moderate,High


# Executive Summary

## Dataset Overview

- Total Employees: **1,470**
- Features: **35**
- Numerical Features: **26**
- Categorical Features: **9**
- Missing Values: **0**
- Duplicate Records: **0**

---

## Data Quality

The dataset is complete, consistent, and suitable for machine learning.

Three constant-value features (EmployeeCount, Over18, StandardHours) and one identifier (EmployeeNumber) were identified for removal during preprocessing.

---

## Key Business Findings

- Approximately **16%** of employees left the organization.
- Overtime is strongly associated with employee attrition.
- Employees with lower job satisfaction exhibit higher attrition rates.
- Poor work-life balance is linked with increased turnover.
- Monthly income and total working experience appear related to employee retention.
- Business travel may contribute to higher attrition for some employee groups.

---

## Recommended Preprocessing Steps

- Remove constant-value features.
- Remove EmployeeNumber.
- Encode categorical variables using One-Hot Encoding.
- Scale numerical variables where required.
- Preserve naturally occurring outliers unless they negatively affect model performance.
- Evaluate class imbalance during model experimentation.

---

